### DISCUSSION TITLE
Can't use Tor SOCKS proxy but HTTP proxy works fine: TypeError: […] 'max_connections'

### DISCUSSION THREAD

**Discussion by Nerdmind** [searx/searx#3164] [Installation issue: custom]
**Version**
The `master` branch at commit 0248777f959ba9ee26fc8f1c367ea4ef2cb5d7cc

**How did you install Searx?**
https://searx.github.io/searx/admin/installation-searx.html#installation-basic

**What happened?**
When trying to use the Tor network via SOCKS proxy for outgoing search engine traffic, Searx throws the following error messages multiple times for a bunch of engines some seconds after startup and any search request will fail (`unexpected crash`):

~~~
ERROR:searx.engines:wikidata engine: Fail to initialize
Traceback (most recent call last):
  File "/usr/local/searx/src/searx/engines/__init__.py", line 295, in engine_init
    init_fn(get_engine_from_settings(engine_name))
  File "/usr/local/searx/src/searx/engines/wikidata.py", line 689, in init
    jsonresponse = send_wikidata_query(query)
  File "/usr/local/searx/src/searx/engines/wikidata.py", line 134, in send_wikidata_query
    http_response = get(SPARQL_ENDPOINT_URL + '?' + urlencode({'query': query}), headers=get_headers())
  File "/usr/local/searx/src/searx/network/__init__.py", line 105, in get
    return request('get', url, **kwargs)
  File "/usr/local/searx/src/searx/network/__init__.py", line 82, in request
    response = future.result(timeout)
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/local/searx/src/searx/network/network.py", line 201, in request
    client = await self.get_client(**kwargs_clients)
  File "/usr/local/searx/src/searx/network/network.py", line 165, in get_client
    max_redirects
  File "/usr/local/searx/src/searx/network/client.py", line 184, in new_client
    retries)
  File "/usr/local/searx/src/searx/network/client.py", line 146, in get_transport_for_socks_proxy
    **TRANSPORT_KWARGS)
  File "/usr/local/searx/env/lib/python3.7/site-packages/httpx_socks/_async_transport.py", line 63, in __init__
    **kwargs,
TypeError: type object got multiple values for keyword argument 'max_connections'
~~~

**How To Reproduce**
I have the following proxy configuration in the `outgoing` section of my `settings.yml`  which inherits the default settings, and I can confirm that the Tor SOCKS proxy is listening on `127.0.0.1:9050` and works just fine with programs like `curl`.

~~~yaml
outgoing:
    proxies :
        http  : 'socks5h://127.0.0.1:9050'
        https : 'socks5h://127.0.0.1:9050'
~~~

What *does* work without a problem is using the Tor network via HTTP proxy instead of SOCKS proxy with the following configuration (you would need to set `HTTPTunnelPort 9060` in the `torrc`). Then the traffic is definitely flowing through the Tor network.

~~~yaml
outgoing:
    proxies :
        http  : 'http://127.0.0.1:9060'
        https : 'http://127.0.0.1:9060'
~~~

**Expected behavior**
Unless I didn't misunderstood or misconfigured something, it should be possible to use the Tor network via SOCKS proxy?

**Screenshots & Logs**
1. [Full log output after startup](https://github.com/searx/searx/files/8079328/searx.log)
2. [The installed `pip` packages](https://github.com/searx/searx/files/8079646/searx-env-packages.txt)
3. **The web interface after searching for something:**
    ![searx-unexpected-crash](https://user-images.githubusercontent.com/7452174/154257294-6e73429d-9b0e-4741-8e3a-32ccf4ccb34d.png)

**Comment by unixfox** [comment]
Can you post your full settings.yml without any confidential info?

**Reply by Nerdmind** [reply]
Hello. [Here is the complete `settings.yml` file](https://github.com/searx/searx/files/8079868/searx-settings.yml.txt) using the Tor HTTP proxy.

**Reply by unixfox** [reply]
The proxies are not configured correctly, please read the example here: https://github.com/searx/searx/blob/master/searx/settings.yml#L81

**Reply by Nerdmind** [reply]
I can confirm that the syntax I'm using is working at least for the HTTP tunnel, otherwise the traffic wouldn't go through the Tor network? I also saw the other syntax in the link you send and tested but it doesn't make any difference. The behavior is exactly the same.

I also did `pip install requests[socks]` and I just tested it again with exactly the following syntax:

~~~yaml
outgoing:
    […]
    proxies:
        all://:
            - socks5h://127.0.0.1:9050
~~~

It doesn't work. Same error messages showing up. If I change it to use the HTTP tunnel, it works:
~~~yaml
outgoing:
    […]
    proxies:
        all://:
            - http://127.0.0.1:9060
~~~

Setting `using_tor_proxy: True` does make no difference when trying to use SOCKS except that now the web server does not keeps running after executing `python searx/webapp.py` and the shell prompt is returned immediately after the `max_connnections` errors are displayed.

When using the HTTP tunnel and `using_tor_proxy: True`, the following message shows up:  
`Network configuration problem: not using Tor`

The only thing that works is specifying `HTTPTunnelPort 9060` in the `torrc` and using it via `http://127.0.0.1:9060`.

**Reply by Nerdmind** [reply]
OK, after randomly checking out some older commits (and upgrading/downgrading the project modules with `pip install -e .` after each `git checkout`) I finally found something!

The last commit that works with the `socks5h://127.0.0.1:9050` configuration line (and **definitely** sends the traffic through the Tor network) is 321ddc91bca5e2267e2fb0ab7233cc403f5a456f. So this means the next commit 3ddd0f8944c98cddbc136d73adc25e4399a5cb18 has introduced the problem.

I hope that helps you developers figure out what exactly the problem is. I'm going to use the HTTP tunnel until then.

Thank you! :smile:

**Reply by unixfox** [reply]
Could you try if it works on this popular fork of searx: https://github.com/searxng/searxng and could report this bug to searxng: https://github.com/searxng/searxng/issues?

The dev that worked a lot with httpx at searx left the project for searxng so you might have a fix for your issue quickly on searxng.

**Reply by Nerdmind** [reply]
I just checked out SearxNG and it seems to be not affected. It works with both HTTP and SOCKS proxy.

**Reply by unixfox** [reply]
Awesome, searx is quite outdated so that's why searx could have some issues related to socks proxies.

**Comment by MarcAbonce** [comment]
@Nerdmind, thank you for reporting this bug and pointing out the exact commit that caused the issue.

I think #3165 should fix it, but you can try those changes locally if you want to confirm.

**Reply by Nerdmind** [reply]
Hello. I checked out your branch and it seems to work when using this configuration:

~~~yaml
outgoing:
    proxies :
        all://: 'socks5h://localhost:9050'
~~~

The `using_tor_proxy` setting inside the `outgoing` section does also seems to work, at least when I remove the `proxies` section, the `Network configuration problem: not using Tor` message is logged.
